python/telemetry/tubix20.py

```diff
@@ -0,0 +1,71 @@
+#!/usr/bin/env python3
+# -*- coding: utf-8 -*-
+
+# Copyright 2025 Fabian P. Schmidt <kerel@mailbox.org>
+#
+# This file is part of gr-satellites
+#
+# SPDX-License-Identifier: GPL-3.0-or-later
+#
+"""
+TUBiX20 Telemetry Format
+
+# References
+[1]: https://www.static.tu.berlin/fileadmin/www/10002275/
+     Amateur_Radio/TechnoSat_Telemetry_Format.ods
+"""
+
+from construct import *
+
+
+MessageType = Enum(
+    BitsInteger(3),
+    ack=0,
+    regular=1,
+    error_correction=7)
+
+TargetAddress = Enum(
+    BitsInteger(4),
+    gs=0,
+    broadcast=15,
+)
+
+TargetSubAddress = Enum(
+    BitsInteger(2),
+    primary_gs=0,
+    secondary_gs=1,
+    broadcast=3,
+)
+
+Baudrate = Enum(
+    BitsInteger(1),
+    baud_4k8=0,
+    baud_9k6=1,
+)
+
+Control = BitStruct(
+    'message_type' / MessageType,
+    'num_blocks' / BitsInteger(5),
+    'address' / TargetAddress,
+    'sub_address' / TargetSubAddress,
+    'ack' / Flag,
+    'baud' / Baudrate,
+    )
+
+TransferFrameHeader = BitStruct(
+    'scid' / BitsInteger(12),
+    'version' / BitsInteger(4),
+    'counter' / BitsInteger(16),
+)
+
+MasterFrame = Struct(
+    'control' / Control,
+    'callsign' / Bytes(6),
+    'callsign_crc' / Hex(Int16ub),
+    'transfer_frame' / Bytes(lambda ctx: 18 * (ctx.control.num_blocks + 1)),
```

Comment:
I think this `lambda` can be written as `18 * (this.control.num_blocks + 1)`, but I've also found some pitfalls with `this` in `construct`, so I don't have a preference for using `this` or an explicit `lambda`.

Reply:
The pitfalls are that `this` is magic object that you can do operations with to construct a `lambda` in a shorthand way. However those operations are limited to what is allowed by Python's operator overloading. For instance, `18 * (this.control.num_blocks + 1)` should work because `__getattr__()` and the `+` and `*` operators are overloaded to do the required magic. However something like `if this == 1` doesn't work, because it's impossible to overload what it means for an object to be casted to a `bool` as part of a conditional expresssion.